File: backend/internal/model/glossary.go
package model

// GlossaryEntry is one normalized term: source (usually Japanese) -> translation
// (usually Chinese), plus optional aliases and a free-form note. Category is the
// originating sheet name (e.g. "专有名词表"); SubCategory is the in-sheet section
// header (e.g. "人称", "音乐类"). Origin tracks where the entry came from so that
// re-import / remote sync can replace import|remote rows while preserving the
// user's hand-added rows.
type GlossaryEntry struct {
	ID          string   `json:"id"`
	Source      string   `json:"source"`
	Translation string   `json:"translation"`
	Aliases     []string `json:"aliases,omitempty"`
	Note        string   `json:"note,omitempty"`
	Category    string   `json:"category"`
	SubCategory string   `json:"subCategory,omitempty"`
	Origin      string   `json:"origin"` // "import" | "user" | "remote"

	// Team-mode collaboration fields, populated from a remote /export sync.
	// All omitempty so local-only libraries keep their existing JSON shape.
	ContributorName string `json:"contributorName,omitempty"`
	UpdatedBy       string `json:"updatedBy,omitempty"`
	Version         int    `json:"version,omitempty"`
}

// Origin values.
const (
	OriginImport = "import"
	OriginUser   = "user"
	OriginRemote = "remote"
)

// Appellation is one cell of the 人称表 matrix: how Speaker addresses Target,
// in Japanese (JP) and Chinese (CN). Not searched in v1 — surfaced through the
// dedicated double-dropdown lookup ("A 称呼 B 为 ...").
type Appellation struct {
	Speaker string `json:"speaker"`
	Target  string `json:"target"`
	JP      string `json:"jp,omitempty"`
	CN      string `json:"cn,omitempty"`
}

// GrammarUsage is one row of the 语法用例 sheet: a grammar point with its
// connection pattern, an example sentence, and provenance. Surfaced through the
// dedicated Grammar page, separate from the term glossary.
type GrammarUsage struct {
	ID         string `json:"id"`
	Item       string `json:"item"`                 // 语法项目
	Location   string `json:"location,omitempty"`   // 出现位置
	Index      string `json:"index,omitempty"`      // 例句标号
	Connection string `json:"connection,omitempty"` // 接续
	Note       string `json:"note,omitempty"`       // 备注
	Unit       string `json:"unit,omitempty"`       // 来自哪个团
	Example    string `json:"example,omitempty"`    // 例句（含溢出列拼接）
	Reference  string `json:"reference,omitempty"`  // 参考翻译
}

// GlossaryData is the full persisted payload.
type GlossaryData struct {
	Entries      []GlossaryEntry `json:"entries"`
	Appellations []Appellation   `json:"appellations"`
	Grammar      []GrammarUsage  `json:"grammar,omitempty"`
}

File: src/types/glossary.ts
export interface GlossaryEntry {
  id: string
  source: string
  translation: string
  aliases?: string[]
  note?: string
  category: string
  subCategory?: string
  origin: 'import' | 'user' | 'remote'
  // Team-mode collaboration fields (populated from a remote sync).
  contributorName?: string
  updatedBy?: string
  version?: number
}

// --- Team mode (collaborative glossary server) ---

export interface TeamUser {
  id: string
  username: string
  displayName: string
  role: 'member' | 'reviewer' | 'admin' | 'superadmin'
  status: string
  // User-chosen avatar background colour ('#rrggbb'); empty/undefined → a
  // deterministic colour derived from the user id (see AccountPage).
  avatarColor?: string
}

export interface TeamStatus {
  loggedIn: boolean
  connected: boolean
  readonly: boolean
  serverUrl: string
  user: TeamUser | null
}

export interface Proposal {
  id: string
  kind: 'add' | 'edit' | 'delete'
  targetType: string
  targetId?: string
  category: string
  payload: string
  baseVersion?: number
  authorId: string
  authorName?: string
  status: 'pending' | 'approved' | 'rejected'
  reviewerId?: string
  reviewNote?: string
  createdAt: number
  reviewedAt?: number
}

export interface CategoryCount {
  category: string
  count: number
}

export interface Appellation {
  speaker: string
  target: string
  jp?: string
  cn?: string
}

export interface AppellationResult {
  found: boolean
  speaker?: string
  target?: string
  jp?: string
  cn?: string
}

// Per-sheet outcome of an Excel import.
export interface SheetReport {
  sheet: string
  kind: 'terms' | 'appellations' | 'grammar' | 'skipped'
  count: number
  skipped: string
}

export interface ImportReport {
  sheets: SheetReport[]
  totalEntries: number
  totalAppellations: number
  totalGrammar: number
}

// One grammar usage row (语法用例 sheet), surfaced on the dedicated Grammar page.
export interface GrammarUsage {
  id: string
  item: string
  location?: string
  index?: string
  connection?: string
  note?: string
  unit?: string
  example?: string
  reference?: string
}

// Full payload returned by /glossary/export.
export interface GlossaryData {
  entries: GlossaryEntry[]
  appellations: Appellation[]
  grammar?: GrammarUsage[]
}
